File: creational/simple_factory/simple_factory.go
package simple_factory

type Animal interface {
	Speak() string
}

func NewAnimal(animalType string) Animal {
	switch animalType {
	case "Dog":
		return &Dog{}
	case "Cat":
		return &Cat{}
	default:
		panic ("Invalid type")
	}
}

type Cat struct {
}

func (c *Cat) Speak() string {
	return "Meow!"
}

type Dog struct {
}

func (d *Dog) Speak() string {
	return "Woof!"
}

File: creational/simple_factory/simple_factory.py
class Animal:
    def __init__(self, name: str) -> None:
        self.name = name

    def speak(self):
        pass


class Dog(Animal):
    def speak(self):
        return "Woof!"


class Cat(Animal):
    def speak(self):
        return "Meow!"


class AnimalFactory:
    def create_animal(self, name: str) -> Animal:
        if name == "dog":
            return Dog(name)
        elif name == "cat":
            return Cat(name)
        else:
            return None


def main():
    factory = AnimalFactory()
    dog = factory.create_animal("dog")
    print(dog.speak())

    cat = factory.create_animal("cat")
    print(cat.speak())


if __name__ == "__main__":
    main()
